### PR TITLE
VFB-23 Add List, Create, and Delete User Functionality for Admin Panel

### PR DESCRIPTION
**Pull request by daniel-conlan-softwire**
Implemented a basic administrator panel that provides:
- Table of all users registered with the foodbank.
- Form to create new user accounts for the foodbank.
- Option to delete existing user accounts from the foodbank.

Administrative actions are restricted to admin accounts only - regular user accounts cannot perform these actions.

This is an intermediate pull request and there is still quite a lot of work still required to tidy the UI. As such, there will be a lot of `TODO` comments littered throughout the PR. These comments will be addressed in a future PR. Some functionality may be partially implemented, and there is likely to be some small bugs that shall be squashed shortly.

- [X] Make sure you've verified it works via `npm run dev`
- [X] Make sure you've linted via `npm run lint`
    - can run `npm run lint_fix` to try and fix as any mistakes automatically as possible!
- [X] Make sure you've tested via `npm run test`

### CODE REVIEW

File: src/app/admin/AdminPage.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+`;
+
+interface Props {
+    userData: UserRow[];
+}
+
+// TODO VFB-23 Add accessibility tests for the admin page
+
+const AdminPage: React.FC<Props> = (props) => {
+    const adminPanels: [string, IconDefinition, ReactElement][] = [
```
Comment (TomCusack-Softwire) [inline src/app/admin/AdminPage.tsx:25]: can you interface this and make it a list of objects? I'm not a massive fan of big lists that you only really know what's going on by their type and the usage (relying on the _user_ of the function to use clear variable names!)

File: src/app/admin/AdminPage.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        ["USERS TABLE", faUsers, <UsersTable key={1} userData={props.userData} />],
```
Comment (TomCusack-Softwire) [inline src/app/admin/AdminPage.tsx:26]: is `key={1}` required, especially if you have a `key={index}` already on the TableSurface?

File: src/app/admin/AdminPage.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        ["CREATE USER", faUserPlus, <CreateUserForm key={2} />],
+    ];
+
+    return (
+        <>
+            {adminPanels.map(([panelTitle, panelIcon, PanelContent], index) => {
+                return (
+                    <TableSurface key={index}>
+                        <Accordion elevation={0}>
```
Comment (DanWilkins2107) [inline src/app/admin/AdminPage.tsx:35]: What does elevation={0} do? Surely that is default?

File: src/app/admin/RefreshPageButton.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import Button from "@mui/material/Button";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faArrowsRotate } from "@fortawesome/free-solid-svg-icons/faArrowsRotate";
+import React from "react";
+
+const RefreshPageButton: React.FC<{}> = () => {
+    const refreshPage = (): void => {
+        window.location.reload();
```
Comment (TomCusack-Softwire) [inline src/app/admin/RefreshPageButton.tsx:8]: if at all possible I prefer NextJS specific things rather than window-specific things, in this case
```ts
const router = useRouter(); // from next/navigation
...
onClick={router.refresh}
// OR
onClick={()=>router.refresh()}
```

File: src/app/admin/adminActions.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import { FunctionsResponse } from "@supabase/functions-js/src/types";
+import { User } from "@supabase/gotrue-js";
+import { Datum } from "@/components/Tables/Table";
+
+const supabase = createServerActionClient({ cookies });
+
+export interface UserRow extends Datum {
+    id: string;
+    email: string;
+    userRole: "admin" | "caller";
```
Comment (TomCusack-Softwire) [inline src/app/admin/adminActions.ts:14]: should be from `Database["public"]["Enums"]["gender"]` instead, in case we want to add more rows!

File: src/app/admin/createUser/CreateUserForm.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                {refreshRequired ? (
+                    <RefreshPageButton />
+                ) : (
+                    <Button
+                        startIcon={<FontAwesomeIcon icon={faUserPlus} />}
+                        variant="contained"
+                        onClick={submitForm}
+                        disabled={submitDisabled}
+                    >
+                        Create Client
```
Comment (DanWilkins2107) [inline src/app/admin/createUser/CreateUserForm.tsx:97]: Thought this creates a user not a client?

File: src/app/admin/page.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import React, { ReactElement } from "react";
+import Title from "@/components/Title/Title";
+import { Metadata } from "next";
+import AdminPage from "@/app/admin/AdminPage";
+import { getUsers } from "@/app/admin/adminActions";
+
+// disables caching
+export const revalidate = 0;
```
Comment (DanWilkins2107) [inline src/app/admin/page.tsx:8]: How come we need to disable caching in this case? Not suggesting it isn't needed, just don't fully understand why we do this.

File: src/app/admin/createUser/CreateUserForm.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            setSubmitError(Errors.submit);
+            setSubmitDisabled(false);
+            return;
+        }
+
+        const response = await createUser(fields);
+
+        if (response.error) {
+            setSubmitError(Errors.external);
+            setSubmitErrorMessage(response.error.message);
```
Comment (TomCusack-Softwire) [inline src/app/admin/createUser/CreateUserForm.tsx:64]: do you unset the submiterror on success?

File: src/app/admin/usersTable/UsersTable.tsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+export const OptionButtonDiv = styled.div`
+    display: flex;
+    padding-top: 1rem;
+    gap: 1rem;
+    justify-content: center;
+`;
+
+const usersTableHeaderKeysAndLabels: TableHeaders = [
+    ["id", "USER ID"],
```
Comment (TomCusack-Softwire) [inline src/app/admin/usersTable/UsersTable.tsx:26]: please 'normal case' the headers and set capitalisation in CSS (screen readers do much better with this!)

File: src/app/admin/usersTable/UsersTable.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+            <DangerDialog
+                header="DELETE USER"
+                headerId="deleteUserDialog"
+                isOpen={userToDelete !== undefined}
+                onClose={onUserDeleteCancellation}
+            >
+                Are you sure you want to delete user <b>{userToDelete ? userToDelete.email : ""}</b>
+                ?
+                <br />
```
Comment (DanWilkins2107) [inline src/app/admin/usersTable/UsersTable.tsx:103]: What does the `<br />` do that is better than flexboxing and using gap?

File: src/components/DataInput/DataInput.cy.tsx [hunk 11]
```diff
@@ -40,6 +41,13 @@ describe("Data Input Components", () => {
                 onChange={getFreeFormTextHandler(() => console.log("Freeform Changed"))}
             />
         );
+        cy.mount(
+            <PasswordInput
+                label="Password"
+                defaultValue="password"
+                onChange={getFreeFormTextHandler(() => console.log("Password Chanegd"))}
```
Comment (DanWilkins2107) [inline src/components/DataInput/DataInput.cy.tsx:48]: Changed*

File: src/app/clients/ClientsPage.tsx [hunk 12]
```diff
@@ -20,8 +20,6 @@ import TableSurface from "@/components/Tables/TableSurface";
 import { CenterComponent } from "@/components/Form/formStyling";
 import Button from "@mui/material/Button";
 
-// TODO Change Button to LinkButton
```
Comment (TomCusack-Softwire) [inline src/app/clients/ClientsPage.tsx:23]: why is this removed?

File: src/components/DataInput/DataInput.cy.tsx [hunk 13]
```diff
@@ -217,11 +258,5 @@ describe("Data Input Components", () => {
             cy.get("input[value='c']").click();
             cy.get("@onChangeSpy").should("have.been.calledWith", "c");
         });
-
-        it("Add error message to form", () => {
```
Comment (TomCusack-Softwire) [inline src/components/DataInput/DataInput.cy.tsx:221]: does this no longer work?

File: supabase/functions/admin-create-user/index.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    const SUPABASE_ANON_KEY = Deno.env.get("SUPABASE_SERVICE_ROLE_KEY");
+    const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY, {
+        global: {
+            headers: {
+                Authorization: req.headers.get("Authorization")!,
+            },
+        },
+    });
+
+    // -- Prevent Non-Admin Access -- //
```
Comment (TomCusack-Softwire) [inline supabase/functions/admin-create-user/index.ts:35]: did you say you were going to generalise this bit to the _shared folder?

**Review by TomCusack-Softwire — CHANGES_REQUESTED** [review]
Looks pretty good!

File: src/app/admin/AdminPage.tsx (hunk 3 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/AdminPage.tsx:35]: Sadly, 'tis actually not. The accordion is elevated by default, and I just want it to be flush with the `TableSurface` - looks a bit strange otherwise

File: src/app/admin/adminActions.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+            email: user.email ?? "-",
+            userRole: user.app_metadata.role,
+            createdAt: Date.parse(user.created_at),
+            updatedAt: Date.parse(user.updated_at ?? ""),
+        };
+    });
+};
+
+export const createUser = async (fields: any): Promise<FunctionsResponse<any>> => {
+    return await supabase.functions.invoke("admin-create-user", {
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/adminActions.ts:40]: I think what I'll do is handle the errors within the client components and just return the raw response from the edge functions

File: src/app/admin/createUser/CreateUserForm.tsx (hunk 6 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/createUser/CreateUserForm.tsx:97]: Yep, good spot

File: src/app/admin/createUser/createUserFormSections/AccountDetails.tsx [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+import FreeFormTextInput from "@/components/DataInput/FreeFormTextInput";
+import PasswordInput from "@/components/DataInput/PasswordInput";
+import GenericFormCard from "@/components/Form/GenericFormCard";
+import { CardProps, errorExists, errorText, onChangeText } from "@/components/Form/formFunctions";
+
+const AccountDetails: React.FC<CardProps> = ({ fieldSetter, formErrors, errorSetter }) => {
+    return (
+        <GenericFormCard
+            title="Account Details"
+            required={true}
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/createUser/createUserFormSections/AccountDetails.tsx:11]: Changed

File: src/app/admin/page.tsx (hunk 7 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/page.tsx:8]: Emm, I believe we were having some problems with results from calls to the edge functions caching things that shouldn't be cached - I'll have more of a look into this later, it's on my `TODO` list

File: src/app/admin/usersTable/UsersTable.tsx (hunk 10 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/usersTable/UsersTable.tsx:103]: A dinnae ken, really. I guess it was just easier for me to do it this way. For just now I think this is okay, but I'll keep it in mind for later

File: src/components/DataInput/DataInput.cy.tsx (hunk 11 above)
Comment (daniel-conlan-softwire) [reply src/components/DataInput/DataInput.cy.tsx:48]: This actually isn't a mistake, it's derived from the old latin for the infinitive "to change"

...no I'm just kidding, that's a good spot

File: src/app/admin/AdminPage.tsx (hunk 2 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/AdminPage.tsx:26]: I thought the same thing, but I believe the linter or something was complaining because these objects were in a list. I'll see if it changes after fixing the above

File: src/app/admin/AdminPage.tsx (hunk 1 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/AdminPage.tsx:25]: I've changed it now

File: src/app/admin/RefreshPageButton.tsx (hunk 4 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/RefreshPageButton.tsx:8]: I tried this previously, but I couldn't quite get it to work. The page doesn't appear to refresh when I try any of the above methods

File: src/app/admin/adminActions.ts [hunk 17]
```diff
@@ -0,0 +1,49 @@
+"use server";
+
+import { createServerActionClient } from "@supabase/auth-helpers-nextjs";
+import { cookies } from "next/headers";
+import { FunctionsResponse } from "@supabase/functions-js/src/types";
+import { User } from "@supabase/gotrue-js";
+import { Datum } from "@/components/Tables/Table";
+
+const supabase = createServerActionClient({ cookies });
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/adminActions.ts:9]: So....server actions are a relatively new thing for Next.JS that automatically create an API endpoint that can be called from client components without having to define the endpoint manually. So, every function inside of the `adminActions.ts` file have been defined as server actions and has its own endpoint within Next

Server components I guess are just, well, components that are processed on the server

Server actions can be called from within a server component as well as a client component

File: src/app/admin/createUser/CreateUserForm.tsx (hunk 8 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/createUser/CreateUserForm.tsx:64]: Nope! I noticed a bug with this previously...and proceeded not to check it out and have it fixed. But it's fixed now - good catch

File: src/app/admin/createUser/createUserFormSections/AccountDetails.tsx [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+            text="Please enter the email and temporary password for the new user."
+        >
+            <>
+                {/* TODO VFB-23 ADD REGEX CHECK FOR EMAIL */}
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/createUser/createUserFormSections/AccountDetails.tsx:15]: Added another `TODO`

File: src/app/admin/usersTable/UsersTable.tsx (hunk 9 above)
Comment (daniel-conlan-softwire) [reply src/app/admin/usersTable/UsersTable.tsx:26]: I've just decapitalised these for just now

File: src/components/NavBar/NavigationBar.tsx [hunk 19]
```diff
@@ -126,6 +126,7 @@ const ResponsiveAppBar: React.FC<Props> = ({ children }) => {
         ["Clients", "/clients"],
         ["Lists", "/lists"],
         ["Calendar", "/calendar"],
+        ["Admin", "/admin"], // TODO REMOVE - DEBUG ONLY
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/components/NavBar/NavigationBar.tsx:129]: I've changed this `TODO` to be more specific as to what needs changed

File: supabase/functions/admin-create-user/index.ts (hunk 14 above)
Comment (daniel-conlan-softwire) [reply supabase/functions/admin-create-user/index.ts:35]: I've added it to my `TODO` list and I'll change it in a separate PR

File: src/app/admin/usersTable/UsersTable.tsx [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+
+const formatTimestamp = (timestamp: number): string => {
+    if (isNaN(timestamp)) {
+        return "-";
+    }
+
+    return new Date(timestamp).toLocaleString("en-gb");
+};
+
+const userTableColumnDisplayFunctions: { [headerKey: string]: ColumnDisplayFunction } = {
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/usersTable/UsersTable.tsx:41]: Changed

File: src/app/clients/ClientsPage.tsx (hunk 12 above)
Comment (daniel-conlan-softwire) [reply src/app/clients/ClientsPage.tsx:23]: I've added it back in

File: src/app/admin/adminActions.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    email: string;
+    userRole: "admin" | "caller";
+    createdAt: number;
+    updatedAt: number;
+}
+
+export const getUsers = async (): Promise<UserRow[]> => {
+    const { data, error } = await supabase.functions.invoke("admin-get-users");
+
+    if (error) {
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/adminActions.ts:22]: I've added a `throw` statement as requested and an additional `TODO` for changing the error handling for a future PR on this ticket

File: src/components/Tables/Table.tsx [hunk 22]
```diff
@@ -15,7 +15,7 @@ import IconButton from "@mui/material/IconButton/IconButton";
 import Icon from "@/components/Icons/Icon";
 
 export interface Datum {
-    [headerKey: string]: string | boolean | null;
+    [headerKey: string]: string | number | boolean | null;
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/components/Tables/Table.tsx]: Done

File: src/app/clients/ClientsPage.tsx [hunk 23]
```diff
@@ -21,6 +21,8 @@ import { CenterComponent } from "@/components/Form/formStyling";
 import ActionBar from "@/app/clients/ActionBar";
 import AddParcelsButton from "@/app/clients/AddParcelsButton";
 
+// TODO Change Button to LinkButton
```
Comment (TomCusack-Softwire) [inline src/app/clients/ClientsPage.tsx:24]: wait this is no longer on main - did you rebase recently? please remove now :) [sorry]

**Review by TomCusack-Softwire — APPROVED** [review]
Just a tiny change!

File: src/app/clients/ClientsPage.tsx (hunk 23 above)
Comment (daniel-conlan-softwire) [reply src/app/clients/ClientsPage.tsx:24]: It's all good, I noticed that when I rebased back onto `main`

**Review by DanWilkins2107 — APPROVED** [review]

File: src/app/admin/createUser/CreateUserForm.tsx [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+} from "@/components/Form/formFunctions";
+import { createUser } from "@/app/admin/adminActions";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faUserPlus } from "@fortawesome/free-solid-svg-icons";
+import RefreshPageButton from "@/app/admin/RefreshPageButton";
+
+interface CreateUserDetails {
+    email: string;
+    password: string;
+    role: "admin" | "caller";
```
[the remark this answers is not in the document]
Comment (daniel-conlan-softwire) [reply src/app/admin/createUser/CreateUserForm.tsx:21]: Resolved
